src/timestamp.rs

```diff
@@ -756,52 +841,58 @@ pub mod context {
 
-                let last_reseed = self.last_reseed.get();
+                let mut last_reseed = self.last_reseed.get();
+                let mut counter;
 
                 // If the observed system time has shifted forwards then regenerate the counter
-                if millis > last_reseed.millis {
-                    let last_reseed = LastReseed::from_millis(millis);
+                if last_reseed.should_reseed(seconds, subsec_nanos) {
+                    last_reseed = LastReseed::from_ts(seconds, subsec_nanos);
                     self.last_reseed.set(last_reseed);
 
-                    let counter = crate::rng::u64() & RESEED_MASK;
+                    counter = crate::rng::u64() & RESEED_MASK;
                     self.counter.set(counter);
-
-                    (counter, last_reseed.ts_seconds, last_reseed.ts_subsec_nanos)
                 }
                 // If the observed system time has not shifted forwards then increment the counter
                 else {
                     // If the incoming timestamp is earlier than the last observed one then
                     // use it instead. This may happen if the system clock jitters, or if the counter
                     // has wrapped and the timestamp is artificially incremented
-                    let millis = ();
-                    let _ = millis;
 
                     // Guaranteed to never overflow u64
-                    let counter = self.counter.get() + 1;
+                    counter = self.counter.get() + 1;
 
                     // If the counter has not overflowed its 42-bit storage then return it
                     if counter <= MAX_COUNTER {
                         self.counter.set(counter);
-
-                        (counter, last_reseed.ts_seconds, last_reseed.ts_subsec_nanos)
                     }
                     // Unlikely: If the counter has overflowed its 42-bit storage then wrap it
                     // and increment the timestamp. Until the observed system time shifts past
                     // this incremented value, all timestamps will use it to maintain monotonicity
                     else {
                         // Increment the timestamp by 1 milli
-                        let last_reseed = LastReseed::from_millis(last_reseed.millis + 1);
+                        last_reseed = last_reseed.increment();
                         self.last_reseed.set(last_reseed);
 
                         // Reseed the counter
-                        let counter = crate::rng::u64() & RESEED_MASK;
+                        counter = crate::rng::u64() & RESEED_MASK;
                         self.counter.set(counter);
-
-                        (counter, last_reseed.ts_seconds, last_reseed.ts_subsec_nanos)
                     }
+                };
+
+                // If we're using additional precision then mask it into the front of the counter
+                if self.additional_precision_bits != 0 {
```

Comment:
We should be able to handle overflow more naturally here by masking into the counter when we fetch it, rather than at the end